github.com/treeverse/lakefs/pyramid TestMultipleConcurrentReads flakes out

E.g.:
```sh
ariels@redqueen:~/dev/lakeFS$ go test --count=20 --race github.com/treeverse/lakefs/pyramid
--- FAIL: TestMultipleConcurrentReads (0.01s)
    tier_fs_test.go:175: 
                Error Trace:    tier_fs_test.go:206
                                                        tier_fs_test.go:175
                                                        asm_amd64.s:1374
                Error:          Not equal: 
                                expected: []byte{0x68, 0x65, 0x6c, 0x6c, 0x6f, 0x20, 0x77, 0x6f, 0x72, 0x6c, 0x64, 0x21}
                                actual  : []byte{}
                            
                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1,3 +1,2 @@
                                -([]uint8) (len=12) {
                                - 00000000  68 65 6c 6c 6f 20 77 6f  72 6c 64 21              |hello world!|
                                +([]uint8) {
                                 }
                Test:           TestMultipleConcurrentReads
FAIL
FAIL    github.com/treeverse/lakefs/pyramid     53.283s
FAIL
```

--- a/pyramid/tier_fs.go
+++ b/pyramid/tier_fs.go
@@ -239,15 +239,14 @@ func (tfs *TierFS) openFile(fileRef localFileRef, fh *os.File) (*ROFile, error) 
 func (tfs *TierFS) openWithLock(fileRef localFileRef) (*os.File, error) {
 	_, err := tfs.keyLock.Compute(fileRef.filename, func() (interface{}, error) {
 		// check again file existence, now that we have the lock
-		fh, err := os.Open(fileRef.fullPath)
+		_, err := os.Stat(fileRef.fullPath)
 		if err == nil {
 			// file exists after all
 			cacheAccess.WithLabelValues(tfs.fsName, "Hit").Inc()
-			_ = fh.Close()
 			return nil, nil
 		}
 		if !os.IsNotExist(err) {
-			return nil, fmt.Errorf("open file: %w", err)
+			return nil, fmt.Errorf("stat file: %w", err)
 		}
 
 		// get the file from the block storage
@@ -266,7 +265,7 @@ func (tfs *TierFS) openWithLock(fileRef localFileRef) (*os.File, error) {
 
 		written, err := io.Copy(writer, reader)
 		if err != nil {
-			return nil, fmt.Errorf("copying date to file: %w", err)
+			return nil, fmt.Errorf("copying data to file: %w", err)
 		}
 		downloadHistograms.WithLabelValues(tfs.fsName).Observe(float64(written))
 

--- a/pyramid/tier_fs.go
+++ b/pyramid/tier_fs.go
@@ -190,6 +190,7 @@ func (tfs *TierFS) Open(namespace, filename string) (File, error) {
 		return nil, err
 	}
 
+	// check if file is there - without taking the lock
 	fileRef := tfs.newLocalFileRef(namespace, filename)
 	fh, err := os.Open(fileRef.fullPath)
 	if err == nil {
@@ -201,7 +202,7 @@ func (tfs *TierFS) Open(namespace, filename string) (File, error) {
 	}
 
 	cacheAccess.WithLabelValues(tfs.fsName, "Miss").Inc()
-	fh, err = tfs.readFromBlockStorage(fileRef)
+	fh, err = tfs.openWithLock(fileRef)
 	if err != nil {
 		return nil, err
 	}
@@ -232,19 +233,33 @@ func (tfs *TierFS) openFile(fileRef localFileRef, fh *os.File) (*ROFile, error) 
 	}, nil
 }
 
-// readFromBlockStorage reads the referenced file from the block storage
+// openWithLock reads the referenced file from the block storage
 // and places it in the local FS for further reading.
 // It returns a file handle to the local file.
-func (tfs *TierFS) readFromBlockStorage(fileRef localFileRef) (*os.File, error) {
+func (tfs *TierFS) openWithLock(fileRef localFileRef) (*os.File, error) {
 	_, err := tfs.keyLock.Compute(fileRef.filename, func() (interface{}, error) {
-		var err error
+		// check again file existence, now that we have the lock
+		fh, err := os.Open(fileRef.fullPath)
+		if err == nil {
+			// file exists after all
+			cacheAccess.WithLabelValues(tfs.fsName, "Hit").Inc()
+			_ = fh.Close()
+			return nil, nil
+		}
+		if !os.IsNotExist(err) {
+			return nil, fmt.Errorf("open file: %w", err)
+		}
+
+		// get the file from the block storage
 		reader, err := tfs.adapter.Get(tfs.objPointer(fileRef.namespace, fileRef.filename), 0)
 		if err != nil {
 			return nil, fmt.Errorf("read from block storage: %w", err)
 		}
 		defer reader.Close()
 
-		writer, err := tfs.syncDir.createFile(fileRef.fullPath)
+		// write to temp file - otherwise the file is available to other readers with partial data
+		tmpFullPath := fileRef.fullPath + ".tmp"
+		writer, err := tfs.syncDir.createFile(tmpFullPath)
 		if err != nil {
 			return nil, fmt.Errorf("creating file: %w", err)
 		}
@@ -253,12 +268,18 @@ func (tfs *TierFS) readFromBlockStorage(fileRef localFileRef) (*os.File, error) 
 		if err != nil {
 			return nil, fmt.Errorf("copying date to file: %w", err)
 		}
+		downloadHistograms.WithLabelValues(tfs.fsName).Observe(float64(written))
 
 		if err = writer.Close(); err != nil {
-			err = fmt.Errorf("writer close: %w", err)
+			return nil, fmt.Errorf("writer close: %w", err)
 		}
-		downloadHistograms.WithLabelValues(tfs.fsName).Observe(float64(written))
-		return nil, err
+
+		// copy from temp path to actual path
+		if err = tfs.syncDir.renameFile(tmpFullPath, fileRef.fullPath); err != nil {
+			return nil, fmt.Errorf("rename temp file: %w", err)
+		}
+
+		return nil, nil
 	})
 
 	if err != nil {

